Remove bypass for RemoteDOMWindow $vm property access

ruthvikkonda opened the pull request:
#### 4ad3bb8b572370d20751a00b43c4922ee8071123
<pre>
Remove bypass for RemoteDOMWindow $vm property access
<a href="https://bugs.webkit.org/show_bug.cgi?id=300773">https://bugs.webkit.org/show_bug.cgi?id=300773</a>
<a href="https://rdar.apple.com/115751655">rdar://115751655</a>

Reviewed by Yusuke Suzuki and Keith Miller.

While <a href="https://commits.webkit.org/277560@main">https://commits.webkit.org/277560@main</a> fixed the majority of RemoteDOMWindow property access
security checks, $vm still has a hack/bypass. This PR removes that bypass and allows $vm to go thru
the same correct security checks that other properties do.

exposeDollarVM() currently uses hasOwnProperty() to check if $vm has already been exposed. With the
changes in this PR, this causes a SecurityError and a resulting assert because RemoteDOMWindow
is not allowed to access $vm. To solve this, we replace this check with getOwnPropertySlot().

We adjust the existing cross origin property access test to include $vm to ensure we don&apos;t regress
any existing behavior. Lastly, we add a new SI test to ensure that with SI on, all RemoteDOMWindow property
access security checks go thru correctly (including $vm).

Test: http/tests/site-isolation/remotedomwindow-property-access-security-checks.html
* LayoutTests/http/tests/security/cross-origin-window-property-access-expected.txt:
* LayoutTests/http/tests/security/cross-origin-window-property-access.html:
* LayoutTests/http/tests/site-isolation/remotedomwindow-property-access-security-checks-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/remotedomwindow-property-access-security-checks.html: Added.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::exposeDollarVM):
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::JSDOMWindow::getOwnPropertySlot):

Canonical link: <a href="https://commits.webkit.org/301696@main">https://commits.webkit.org/301696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d02d99641e9a860cc4a38fe76ff3058cb3819ceb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126517 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133406 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78192 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2c322115-879e-4637-b32a-413b785e086d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128388 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46796 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54700 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96274 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64379 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7ac10544-0a04-4069-9cbb-b03d98f476f2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129465 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37428 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113145 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76748 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/06bc4dce-c118-4583-b4ef-934c60a477f0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36320 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31323 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76714 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118567 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107236 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31604 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135954 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124982 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40913 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104781 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53694 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104483 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26694 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49961 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28292 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50616 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53128 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58941 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158027 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52410 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39540 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55744 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54145 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
```diff
@@ -198,10 +198,6 @@ bool JSDOMWindow::getOwnPropertySlot(JSObject* object, JSGlobalObject* lexicalGl
 
     ASSERT(lexicalGlobalObject->vm().currentThreadIsHoldingAPILock());
 
-    // FIXME (rdar://115751655): This should be replaced with a same-origin check between the active and target document.
-    if (!is<LocalDOMWindow>(thisObject->wrapped()) && propertyName == "$vm"_s) [[unlikely]]
-        return true;
```
I don't understand, what tests are relying on this test? It seems like it would be somewhat unusual (a cursory grep of LayoutTests indicates `$vm` is accessed via a global variable, `globalThis` or `window`) to access `$vm` via another window/globalObject. 

Worst case there's only 84 tests that use `$vm` anyway so it seems like we should just fix the tests. Code changes here seem more intrusive long term.

ruthvikkonda replied:
I don't know why the exception was even there in the first place. I don't think any tests even invoke $vm for a RemoteDOMWindow.

The method table hook is needed because currently even for RemoteDOMWindow, exposeDollarVM() gets called which is incorrect. RemoteDOMWindow implies cross site which implies cross origin which means that it can't access $vm. So this PR only exposes $vm for LocalDOMWindow.

With my changes, during exposeDollarVM() for RemoteDOMWindow init, when we do if (hasOwnProperty(this, vm.propertyNames->builtinNames().dollarVMPrivateName())), a SecurityError will get thrown because $vm is not allowed for RemoteDOMWindow. Then an assert will get triggered as a result. The PR's change not only prevents this but also makes sense to add logically because $vm should only be exposed for LocalDOMWindow.

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -3069,7 +3069,8 @@ DEFINE_VISIT_CHILDREN_WITH_MODIFIER(JS_EXPORT_PRIVATE, JSGlobalObject);
 {
     RELEASE_ASSERT(g_jscConfig.restrictedOptionsEnabled && Options::useDollarVM());
-    if (hasOwnProperty(this, vm.propertyNames->builtinNames().dollarVMPrivateName()))
+    PropertySlot slot(this, PropertySlot::InternalMethodType::VMInquiry, &vm);
```
Nit: might be worth a comment that we use VMInquiry because of Security checks in DOMWindow.

kmiller68 reviewed (approved):
r=me too.

ruthvikkonda committed:
Remove bypass for RemoteDOMWindow $vm property access
https://bugs.webkit.org/show_bug.cgi?id=300773
rdar://115751655

Reviewed by Yusuke Suzuki and Keith Miller.

While https://commits.webkit.org/277560@main fixed the majority of RemoteDOMWindow property access
security checks, $vm still has a hack/bypass. This PR removes that bypass and allows $vm to go thru
the same correct security checks that other properties do.

exposeDollarVM() currently uses hasOwnProperty() to check if $vm has already been exposed. With the
changes in this PR, this causes a SecurityError and a resulting assert because RemoteDOMWindow
is not allowed to access $vm. To solve this, we replace this check with getOwnPropertySlot().

We adjust the existing cross origin property access test to include $vm to ensure we don't regress
any existing behavior. Lastly, we add a new SI test to ensure that with SI on, all RemoteDOMWindow property
access security checks go thru correctly (including $vm).

Test: http/tests/site-isolation/remotedomwindow-property-access-security-checks.html
* LayoutTests/http/tests/security/cross-origin-window-property-access-expected.txt:
* LayoutTests/http/tests/security/cross-origin-window-property-access.html:
* LayoutTests/http/tests/site-isolation/remotedomwindow-property-access-security-checks-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/remotedomwindow-property-access-security-checks.html: Added.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::exposeDollarVM):
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::JSDOMWindow::getOwnPropertySlot):

Canonical link: https://commits.webkit.org/301696@main
